src/Plutarch/Extra/Time.hs

```diff
@@ -0,0 +1,164 @@
+{-# LANGUAGE OverloadedRecordDot #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Plutarch.Extra.Time (
+    PCurrentTime (..),
+    pcurrentTime,
+    currentTime,
+    passertCurrentTime,
+    pisWithinCurrentTime,
+    pisCurrentTimeWithin,
+) where
+
+import Control.Composition ((.*))
+import GHC.Records (HasField)
+import Plutarch.Api.V1 (
+    PExtended (PFinite),
+    PInterval (PInterval),
+    PLowerBound (PLowerBound),
+    PPOSIXTime,
+    PUpperBound (PUpperBound),
+ )
+import Plutarch.Api.V2 (PPOSIXTimeRange)
+import Plutarch.Extra.Applicative (pliftA2)
+import Plutarch.Extra.Field (pletAll, pletAllC)
+import Plutarch.Extra.Maybe (passertPJust, pjust, pnothing)
+import Plutarch.Extra.TermCont (pmatchC)
+
+{- | Represent a fully bounded time range.
+
+     Note: 'PCurrentTime' doesn't need a Haskell-level equivalent because it
+     is only used in scripts, and does not go in datums. It is also scott-encoded
+     which is more efficient in usage.
+
+     @since 3.3.0
+-}
+data PCurrentTime (s :: S)
+    = PCurrentTime
+        (Term s PPOSIXTime)
+        -- ^ The lower bound.
+        (Term s PPOSIXTime)
+        -- ^ The upper bound.
+    deriving stock
+        ( -- | @since 3.3.0
+          Generic
+        )
+    deriving anyclass
+        ( -- | @since 3.3.0
+          PlutusType
+        , -- | @since 3.3.0
+          PEq
+        )
+
+-- | @since 3.3.0
+instance DerivePlutusType PCurrentTime where
+    type DPTStrat _ = PlutusTypeScott
+
+{- | Get the current time, given a 'PPOSIXTimeRange'.
+
+     If it's impossible to get a fully-bounded time, (e.g. either end of the 'PPOSIXTimeRange' is
+     an infinity) then we return nothing.
```

Comment:
"If it's impossible to get a full-bounded time (for example, either end of the 'PPOSIXTimeRange' is an infinity), then we return 'PNothing'."